### ISSUE TITLE
Add rule to check that parameters with defaults declared last

### ISSUE THREAD

**Issue by ViLPy** [eslint/eslint#10328]
**Please describe what the rule should do:**

Check that function parameters with default values declared at the end of parameters list

**What category of rule is this? (place an "X" next to just one item)**

[X] Enforces code style
[ ] Warns about a potential error
[ ] Suggests an alternate way of doing something
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

```js
const multiply = (a = 1, b) => a * b;

function reduce(state = initialState, action) {}
```

**Why should this rule be included in ESLint (instead of a plugin)?**

Function parameters with default values which are declared before the function parameters without default values may not take any advantage and could lead to possible misuses.

For example:

```js
const multiply = (a = 1, b) => a * b;

// might lead to
multiply(5); // default parameter misuse, returns NaN as b is undefined
multiply(undefined, 5); // no advantage for default value
```

Also parameters without defaults are usually more essential to the semantics of a function, and provide a stable initial pattern of use where functions are invoked, so that looks like a good common practice for function signatures to place parameters without defaults at the beginning.

**Comment by platinumazure** [comment]
Hi @ViLPy, thanks for the issue.

I like this idea in general, but I worry that it might be considered too opinionated for ESLint core. (But, the fact that default parameters are still overwritten even when there are parameters without default later in the argument list could be a problem, so maybe that might be enough to warrant inclusion in core.)

I've given it a :+1:. Let's see how the rest of the team feels.

**Comment by ViLPy** [comment]
Hi @platinumazure 

Thanks for your feedback. I agree that this could be too opinionated in certain cases. For example, I took my second example with `reduce` from redux, and redux internally checks that you cannot pass an empty action, so it's technically fine to skip default value for action there and have only one default for the first parameter.

But I thought that this might be a nice candidate for core rule not only because of possible misuses. Since parameters without defaults are usually more essential to the semantics of a function, and provide a stable initial pattern of use where functions are invoked, so that looks like a good common practice for function signatures to place parameters without defaults at the beginning of the list.

I'll also update original issue description to reflect this.
